resolve topic: Add `is:resolved` search keyword/filtering support.

akshatdalton opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
Fixes: #18990.

**Testing plan:** <!-- How have you tested? -->

New test cases are added.

timabbott commented on zerver/lib/narrow.py:
```diff
@@ -84,7 +84,7 @@ def narrow_filter(event: Mapping[str, Any]) -> bool:
-            elif operator == "is" and operand in ["starred"]:
+            elif operator == "is" and operand == "starred":
                 if operand not in flags:
                     return False
```
This logic intentionally used the `in` style because that would support additional flags that work like `starred` more easily.

timabbott commented on zerver/lib/topic.py:
```diff
@@ -82,6 +82,11 @@ def topic_match_sa(topic_name: str) -> "ColumnElement[bool]":
     return topic_cond
 
 
+def get_resolved_topic_condition() -> "ColumnElement[bool]":
```
Let's add `_sa` to the name of this function, since it's a SQLAlchemy function and that's the pattern we're using for them here.

timabbott commented on zerver/tests/test_message_fetch.py:
```diff
@@ -2495,6 +2508,21 @@ def test_get_messages_with_only_searching_anchor(self) -> None:
+        cordelia = self.example_user("cordelia")
+
+        self.send_stream_message(cordelia, "Verona", "whatever1")
+        topic_name = RESOLVED_TOPIC_PREFIX + "foo"
```
I'd call this variable `resolved_topic_name` to make the test more readable.

timabbott commented on zerver/tests/test_message_fetch.py:
```diff
@@ -2495,6 +2508,21 @@ def test_get_messages_with_only_searching_anchor(self) -> None:
+        result = self.get_and_check_messages(
+            dict(narrow=orjson.dumps(narrow).decode(), anchor=anchor, num_before=1, num_after=1)
+        )
+        self.assert_length(result["messages"], 1)
```
I'd confirm the ID of the message matches the `anchor`.

akshatdalton commented on zerver/tests/test_message_fetch.py:
```diff
@@ -527,7 +535,7 @@ def test_build_narrow_filter(self) -> None:
         with open(fixtures_path, "rb") as f:
             scenarios = orjson.loads(f.read())
-        self.assertTrue(len(scenarios) == 9)
+        self.assert_length(scenarios, 10)
```
I'm adding a prep commit for: `assertTrue` -> `assert_length`. (https://github.com/akshatdalton/zulip/commit/45d34d75a0a21a6d12572dac8aa2077c4700ca73)

akshatdalton commented:
Following changes I have addressed in the current push:
1. Resolved https://github.com/zulip/zulip/pull/19230#discussion_r668920016 (Removed the commit).
2. Added `_sa` to the function name. https://github.com/zulip/zulip/pull/19230#discussion_r668920984
3. Renamed the variable to `resolved_topic_name`. https://github.com/zulip/zulip/pull/19230#discussion_r668921483
4. Added a test to confirm the ID of the message matches the anchor. https://github.com/zulip/zulip/pull/19230#discussion_r668921982

@timabbott, this is ready for review.

timabbott commented:
I did a few tweaks to the documentation and merged this, thanks @akshatdalton!

I think we'll want to add this to the "recent topics" filters as well, but I'm not sure there's visual space for doing so without the multi-select `dropdown_list_widget` changes that @aryanshridhar was working on.

aryanshridhar commented:
Yup, Adding an extra filter to the "recent topics" will look a bit congested for users having narrow windows. However, I will try to experiment with it and post them on CZO.
